hilti/toolchain/src/compiler/global-optimizer.cc

```diff
@@ -623,6 +630,118 @@ struct TypeVisitor : OptimizerVisitor, visitor::PreOrder<bool, TypeVisitor> {
     }
 };
 
+struct ConstantFoldingVisitor : OptimizerVisitor, visitor::PreOrder<bool, ConstantFoldingVisitor> {
+    std::map<uint64_t, bool> _constants;
+
+    void collect(Node& node) override {
+        _stage = Stage::COLLECT;
+
+        for ( auto i : this->walk(&node) )
+            dispatch(i);
+    }
+
+    bool prune_uses(Node& node) override {
+        _stage = Stage::PRUNE_USES;
+
+        bool any_modification = false;
+
+        while ( true ) {
+            bool modified = false;
+            for ( auto i : this->walk(&node) ) {
+                if ( auto x = dispatch(i) )
+                    modified = modified || *x;
+            }
+
+            if ( ! modified )
+                break;
+
+            any_modification = true;
+        }
+
+        return any_modification;
+    }
+
+    bool operator()(const declaration::GlobalVariable& x, position_t p) {
+        // We only work on boolean constants.
+        if ( ! (x.isConstant() || x.type() == type::Bool()) )
+            return false;
+
+        switch ( _stage ) {
+            case Stage::COLLECT: {
+                const auto& init = x.init();
+                assert(init);
+
+                if ( auto ctor = init.value().tryAs<expression::Ctor>() )
+                    if ( auto bool_ = ctor->ctor().tryAs<ctor::Bool>() )
+                        _constants[p.node.rid()] = bool_->value();
```

Comment:
TODO for the future: Using `rid()` as the index works for now I believe, but it's not very robust as it depends on the the AST not changing. With the new AST code, this can become the canonical ID of the global.